### DISCUSSION TITLE
FastAPI dependencies tare-down breaks with BaseHTTPMiddleware

### DISCUSSION THREAD

**Discussion by MrSalman333** [fastapi/fastapi#8407] [Questions] [answered]
### First Check

- [X] I added a very descriptive title to this issue.
- [X] I used the GitHub search to find a similar issue and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/samuelcolvin/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
import logging

from fastapi import Depends, FastAPI
from fastapi.middleware.base import BaseHTTPMiddleware
from starlette.middleware import Middleware
from starlette.middleware.base import BaseHTTPMiddleware


class CustomMiddleware(BaseHTTPMiddleware):
    async def dispatch(self, request, call_next):
        return await call_next(request)


app = FastAPI(middleware=[Middleware(CustomMiddleware)])


def dependency_a():
    yield "A"
    logging.warning("WILL NOT RUN")


def dependency_b():
    yield "B"
    logging.warning("WILL NOT RUN")


def dependency_c():
    yield "C"
    logging.warning("WILL RUN")


@app.get(
    "/issue",
    dependencies=[
        Depends(dependency_a),
    ],
)
def show_dependencies_issue(
    b=Depends(dependency_b),
    c=Depends(dependency_c),
):
    return {"status": "ok"}
```

### Description

if you use BaseHTTPMiddleware the tare down of decencies "after yield" will not work on all of them ,, only the last one 

### Operating System

Linux

### Operating System Details

docker image

### FastAPI Version

`>= 0.85`

### Python Version

3.11

### Additional Context

this is after https://github.com/encode/starlette/pull/1609

**Comment by MrSalman333** [comment]
the fix will come from upstream starlette

**Comment by Kludex** [comment]
This is already solved in the current Starlette (0.21.0), FastAPI needs to bump it.

**Comment by etienne-monier** [comment]
Hi,
I have the save issue, which is critical.
Is there a workaround ?

Thanks for the help!

**Comment by Kludex** [comment, marked as answer]
This is already solved in FastAPI.

**Comment by etienne-monier** [comment]
Oh, ok. I had 0.85.1 version. This package evolves quite fast!

Thanks for your (fast) help!

Should this issue be closed?
